app/main.py

```diff
@@ -2,5 +1,29 @@
+from functools import wraps
+from typing import Callable, Tuple, Any
 
 
 def cache(func: Callable) -> Callable:
-    # Write your code here
-    pass
+    func_cache = {}
+
+    @wraps(func)
+    def wrapper(*args: Any) -> Any:
+        if args in func_cache:
+            print("Getting from cache")
+            return func_cache[args]
+        else:
+            print("Calculating new result")
+            result = func(*args)
+            func_cache[args] = result
+            return result
+
+    return wrapper
+
+
+@cache
+def long_time_func(base: int, exponent: int, multiplier: int) -> int:
+    return (base ** exponent ** multiplier) % (base * multiplier)
+
+
+@cache
+def long_time_func_2(numbers: Tuple[int, ...], power: int) -> list[int]:
```

Comment:
The function `long_time_func_2` returns a list, but the return type annotation specifies `list[int]`, which is correct for Python 3.9 and above. Ensure that your Python environment supports this syntax, or use `List[int]` with an appropriate import from `typing` for compatibility with earlier versions.